content/guides/end-to-end-testing/auth0-authentication.md

```diff
@@ -111,10 +117,146 @@ require('dotenv').config()
 }
 ```
 
+Note that `auth0_client_secret` is only needed for
+[programmatic login](#Programmatic-Login).
+
 :::
 
 ## Custom Command for Auth0 Authentication
 
+There are two ways you can authenticate to Auth0:
+
+- [Login with `cy.origin()`](#Login-with-cy-origin)
+- [Programmatic Login](#Programmatic-Login)
+
+### Login with [`cy.origin()`](/api/commands/origin)
+
+Next, we'll write a custom command called `loginToAuth0` to perform a login to
+[Auth0](https://auth0.com). This command will use
+[`cy.origin()`](/api/commands/origin) to
+
+1. Navigate to the Auth0 login
+2. Input user credentials
+3. Sign in and redirect back to the
+   [Cypress Real World App](https://github.com/cypress-io/cypress-realworld-app)
+4. Cache the results with [`cy.session()`](/api/commands/session)
+
+```js
+// cypress/support/auth-provider-commands/auth0.ts
+
+Cypress.Commands.add('loginToAuth0', (username: string, password: string) => {
+  const log = Cypress.log({
+    displayName: 'AUTH0 LOGIN',
+    message: [`🔐 Authenticating | ${username}`],
+    // @ts-ignore
+    autoEnd: false,
+  })
+  log.snapshot('before')
+
+  // App landing page redirects to Auth0.
+  cy.visit('/')
+
+  // Login on Auth0 (we'll add caching later).
+  cy.origin(
+    Cypress.env('auth0_domain'),
+    { username, password },
+    ({ username, password }) => {
+      cy.get('input#username').type(username)
+      cy.get('input#password').type(password)
+      cy.contains('button[value=default]', 'Continue').click()
+    }
+  )
+
+  // Ensure Auth0 has redirected us back to the RWA.
+  cy.url().should('equal', 'http://localhost:3000/')
+
+  log.snapshot('after')
+  log.end()
+})
+```
+
+Now, we can use our `loginToAuth0` command in the test. Below is our test to
+login as a user via Auth0 and run a basic sanity check.
+
+<Alert type="success">
+
+The
+[runnable version of this test](https://github.com/cypress-io/cypress-realworld-app/blob/develop/cypress/tests/ui-auth-providers/auth0.spec.ts)
+is in the
+[Cypress Real World App](https://github.com/cypress-io/cypress-realworld-app).
+
+</Alert>
+
+```js
+describe('Auth0', function () {
+  beforeEach(function () {
+    cy.task('db:seed')
+    cy.intercept('POST', '/graphql').as('createBankAccount')
+    cy.loginToAuth0(
+      Cypress.env('auth0_username'),
+      Cypress.env('auth0_password')
+    )
+    cy.visit('/')
+  })
+
+  it('shows onboarding', function () {
+    cy.contains('Get Started').should('be.visible')
+  })
+})
+```
+
+Lastly, we can refactor our login command to take advantage of
+[`cy.session()`](/api/commands/session) to store our logged in user so we don't
+have to reauthenticate before every test.
+
+```js
+Cypress.Commands.add('loginToAuth0', (username: string, password: string) => {
+  const log = Cypress.log({
+    displayName: 'AUTH0 LOGIN',
+    message: [`🔐 Authenticating | ${username}`],
+    // @ts-ignore
```

Comment:
Try it :D

Reply:
Should be fine in this case since it's just a `.js` snippet. As an example it doesn't really add any value, but makes sense in the RWA. I actually think ours typings support `autoEnd` now but could be wrong.